Adding the ability to specify a function to check if micro should retry a failed rpc call

ynori7 opened the pull request:
I've run into the issue that micro is always retrying failed requests even if the error is on the client's side (such as when you provide a bad ID and the service returns a 404). It makes no sense to retry these, so I've added an additional option to specify a checkIfRetriable function, so now you can say something like this:

```
rsp, err := repo.personReadClient.FindPersonByID(
	ctx,
	req,
	microclient.WithRetries(client.DefaultRetries),
	microclient.WithRequestTimeout(client.DefaultRequestTimeout),
	microclient.WithCheckIfRetriable(microclient.Only500Errors),
)
```

By default it uses the AlwaysRetry function which just returns true (and therefore does the same thing which is currently done).

asim commented:
I've spent a little while thinking about this. While I understand the problem, the solution proposed is not something I'm sure of at the moment. There's a long standing issue of needing more fine grained control over the call functionality. While I've added retries and backoff, it feels like every other thing is a little hack on top. Ideally this should be addressed by providing a configurable `CallFunc` and having the current method be the default behaviour. I need to spend a little bit more time thinking about it.

asim commented:
Ok looks better. I can see how this will be useful but let's address a couple things.

RetryFunc is more useful if it looks like BackoffFunc:
`type RetryFunc func(context.Context, Request, int, err error) (bool, error)`

RetryFunc should take the context, request, retry number and error. It should return bool for whether the request should be retried or not but also an error. Returning an error allows the RetryFunc to customise the response back to the caller if need be or indicates whether an error occurred in the retry logic. 

RetryFunc should also be executed inside the Stream method.

ynori7 commented:
I just pushed another change, is that what you had in mind?

asim commented on client/rpc_client.go:
```diff
@@ -400,7 +405,12 @@ func (r *rpcClient) Stream(ctx context.Context, request Request, opts ...CallOpt
+				return nil, err
+			}
+
+			grr = err
```
What error is this using? Why not rsp.err?

asim commented:
Yea looks fine. Just one comment about stream error. 

Little confused by the logic around checking the function response. Idiomatic go method is to check error value.

```
case err := <-ch:
	// if the call succeeded lets bail early
	if err == nil {
		return nil
	}

	retry, rerr := callOpts.Retry(ctx, request, i, err)
        if rerr != nil {
		return rerr
	}

        if !retry {
            return err
        }

	gerr = err
}
```

ynori7 commented:
I adjusted the logic around the returned error from the retry function now and added the comment to the retry.go file to make it clear that returning either true or a non-nil error will result in the call not being retried.

asim commented on client/rpc_client.go:
```diff
@@ -299,6 +299,17 @@ func (r *rpcClient) Call(ctx context.Context, request Request, response interfac
+				return err
+			}
+
+
```
Unnecessary extra space

asim commented on client/rpc_client.go:
```diff
@@ -400,7 +411,17 @@ func (r *rpcClient) Stream(ctx context.Context, request Request, opts ...CallOpt
+				return nil, rsp.err
+			}
+
+			grr = err
```
This should be `grr = rsp.err`

asim commented:
Looks good. Thanks for the contribution.
